### PR TITLE
Allow speculative evaluation

### LINKED ISSUE #406: [worklets] Speculative evaluation
In the "code idempotency" section of the draft spec (https://drafts.css-houdini.org/worklets/#code-idempotency) there's an encouragement to write idempotent code, because UAs might swap globals.

UAs may also wish to speculatively evaluate worklets: for example in the paint worklet, the calculated values for CSS may be available before the concrete object size, so a UA might speculatively evaluate the worklet using an old value for the size, and discard the result if the speculation turns out to be wrong.

It might be nice if the spec explicitly allowed this?

### PR DESCRIPTION
**Pull request by asajeffrey**
Currently, user agents can only invoke worklets with consistent snapshots of the state. This PR allows user agents to invoke worklet scripts with inconsistent state, as long as they discard the result.

The particular instance of this is paint worklets, since the CSS style values are known before the concrete size, so in Servo we speculate that the size hasn't changed (https://github.com/servo/servo/pull/17810).

Fixes #406.

### CODE REVIEW

File: css-paint-api/Overview.bs [hunk 1]
```diff
@@ -461,12 +469,12 @@ interface PaintSize {
...
 into its appropriate stacking level (as defined by the property the CSS property its associated
-with), given its |concreteObjectSize| (<a>concrete object size</a>) it <em>must</em> run the
-following steps:
+with), given |concreteObjectSize| it <em>must</em> run the following steps:
     1. Let |paintFunction| be the <<paint()>> function on the |box| which the user agent wants to
         draw.
 
-    2. If the <a>paint valid flag</a> for the |paintFunction| is <a>paint-valid</a> the user agent
+    2. If the <a>paint valid flag</a> for the |paintFunction| is <a>paint-valid</a>,
+        and the previous invocation had the same |concreteObjectSize|, the user agent
```
Comment (tschneidereit) [inline css-paint-api/Overview.bs]: This implies that there can only ever be one speculative evaluation at a time. If we want to allow speculating multiple frames ahead, this should probably say something about checking for results of a speculative evaluation that match the input state.

File: css-paint-api/Overview.bs [hunk 2]
```diff
@@ -450,6 +451,13 @@ Note: In a future version of the spec, the author could have the ability to spec
 The {{PaintSize}} object represents the size of the image that the author should draw. This is
 the <a>concrete object size</a> given by the user agent.
 
+The <a>draw a paint image</a> function may be speculatively invoked by
+the user agent at any point, with any |concreteObjectSize|. The
+resulting image is discarded.
+
+Note: Although the image is discarded, it may still be cached, so subsequent invocations
+    of <<paint()>> may use the cached image.
+
```
Comment (asajeffrey) [inline css-paint-api/Overview.bs]: I added text to the note: "Note: User agents may use whichever heuristics to speculate a possible future value for |concretObjectSize|, for example speculating that the size remains unchanged." I'm not sure that we want to say more about how to speculate in the normative text.

I replaced "discarded" by "not displayed", which is hopefully less confusing.

File: css-paint-api/Overview.bs (hunk 1 above)
Comment (asajeffrey) [inline css-paint-api/Overview.bs]: All the discussion of the paint valid flag is dubious, perhaps we should just delete it? It's very single-threaded. This is probably a separate issue though.

File: worklets/Overview.bs [hunk 3]
```diff
@@ -119,6 +119,30 @@ The following techniques are used in order to encourage authors to write code in
...
+invoke methods on a class based on the state of the user agent. To
+increase the concurrency between threads, a user agent may invoke a
+method speculatively, before a consistent snapshot of the state is
+known.
+
+In these specifications user agents might invoke methods on a class at
+any time, and with any arguments, not just ones corresponding to
+consistent states of the user agent. The results of such speculative
+evaluations will be discarded, but may be cached for later use, thus
+increasing the concurrency between the user agent and worklet threads.
```
Comment (asajeffrey) [inline worklets/Overview.bs]: Good point, I tried rephrasing it.

File: css-paint-api/Overview.bs (hunk 1 above)
Comment (bfgeek) [inline css-paint-api/Overview.bs]: There is a separate section which allows for the caching inside "invoke a paint callback", see step 10.

File: worklets/Overview.bs [hunk 4]
```diff
@@ -119,6 +119,30 @@ The following techniques are used in order to encourage authors to write code in
 
  - User agents can create and destroy {{WorkletGlobalScope}}s at any time for these specifications.
 
+Speculative Evaluation {#speculative-evaluation}
+------------------------------------------------
+
+Some specifications which use worklets ([[css-paint-api-1]]) may
```
Comment (bfgeek) [inline worklets/Overview.bs]: please reflow to 100 colwidth.

File: css-paint-api/Overview.bs [hunk 5]
```diff
@@ -450,6 +451,16 @@ Note: In a future version of the spec, the author could have the ability to spec
 The {{PaintSize}} object represents the size of the image that the author should draw. This is
 the <a>concrete object size</a> given by the user agent.
 
+The <a>draw a paint image</a> function may be speculatively invoked by
```
Comment (bfgeek) [inline css-paint-api/Overview.bs]: please reflow to 100 colwidth.

File: css-paint-api/Overview.bs (hunk 1 above)
Comment (asajeffrey) [inline css-paint-api/Overview.bs]: @bfgeek yes, this is why I was thinking we could just remove all the stuff about the paint valid flag, since we're already allowing caching inside "invoke a paint callback". Shall I submit a separate PR for that?

Comment (bfgeek) [inline css-paint-api/Overview.bs]: Yeah that sounds good, we still need to describe how the inputProperties invalidate that image on a box, but we can have that in prose vs. as an algorithm.

Comment (asajeffrey) [inline css-paint-api/Overview.bs]: OK, I'll file an issue for this.

File: css-paint-api/Overview.bs (hunk 5 above)
Comment (asajeffrey) [inline css-paint-api/Overview.bs]: Done.

File: worklets/Overview.bs (hunk 4 above)
Comment (asajeffrey) [inline worklets/Overview.bs]: Done.

**Comment by asajeffrey** [comment]
Thanks!
